server/PathServices/SignalRRealtimeDataRepository.cs

```diff
@@ -41,12 +43,36 @@ public SignalRRealtimeDataRepository(IPathDataRepository pathDataRepository, IPa
         /// <returns>A collection of arriving trains.</returns>
         public Task<IEnumerable<RealtimeData>> GetRealtimeData(Station station)
         {
-            return Task.FromResult(this.GetRealtimeData(station, RouteDirection.ToNY).Union(this.GetRealtimeData(station, RouteDirection.ToNJ)).Where(data => data.DataExpiration > DateTime.UtcNow));
+            var allData = this.GetRealtimeData(station, RouteDirection.ToNY).Union(this.GetRealtimeData(station, RouteDirection.ToNJ));
+            var freshData = allData.Where(dataPoint => dataPoint.DataExpiration > DateTime.UtcNow);
+            if (allData.Count() != freshData.Count())
+            {
+                var staledData = allData.Except(freshData);
+                foreach (var staledDataPoint in staledData)
+                    Log.Logger.Here().Warning("Staled data detected for S:{station} R:{route} with timestamp {updatedDataLastUpdated}, force reconnect maybe needed", station, staledDataPoint.Route.DisplayName, staledDataPoint.LastUpdated);
+
+                Log.Logger.Here().Information("Recreating SignalR hubs following staled data detection...");
+                Task.Run(this.CreateHubConnections).Wait();
```

Comment:
If you make this method async you could await CreateHubConnections and just return freshData (rather than needing to wrap it in a task).

Reply:
I think once I deal with the concurrency issue of `CreateHubConnections` I'll avoid calling `.Wait()` so this function can return immediately.
Currently there are 2 potential issues:
```
                    if (!this.hubConnections.ContainsKey((station, direction)))
```
This check is not reliable as without lock another thread can substitute the connection with a new one and the condition will still return true.
```
        private async Task CloseExistingHubConnections()
        {
            // Materialize the connections so we can clear the dictionary before disconnecting.
            // Otherwise, we will reconnect before reinitializing the connection (potentially
            // causing a loop if the token changes).
            var connections = this.hubConnections.Values.ToArray();
            this.hubConnections.Clear();

            await Task.WhenAll(connections.Select(async (client) => await client.DisposeAsync()));
        }
```
Another thread might be able to run between `.ToArray()` and `.Clear()` such that it will try to dispose the connections again.

The goal is to make sure once an `HubConnection` instance is retrieved from dictionary it remain exclusive to the thread until being put back(or not), maybe `TryRemove` will do.

Once refactor is done we should be able to reconnect a single `HubConnection` for more fine-grained control, if staled data is detected for the `(Station, RouteDirection)` pair.